pkg/agent/controller/networkpolicy/priority.go

```diff
@@ -26,125 +26,235 @@ import (
+	priorityIndex := int32(math.Floor(p.PolicyPriority))
+	if priorityIndex > InitialPriorityZones-1 {
+		priorityIndex = InitialPriorityZones - 1
+	}
+	// cannot return a negative OF priority
+	if PriorityTopCNP-InitialPriorityOffset*uint16(priorityIndex) <= uint16(p.RulePriority) {
+		return PriorityBottomCNP
+	}
+	return PriorityTopCNP - InitialPriorityOffset*uint16(priorityIndex) - uint16(p.RulePriority)
+}
+
+// priorityAssigner is a struct that maintains the current mapping between types.Priority and
+// OpenFlow priorities in a single OVS table. It also knows how to re-assign priorities if certain section overflows.
 type priorityAssigner struct {
-	// priorityMap maintains the current mapping between a known CNP priority to OF priority.
+	// priorityMap maintains the current mapping of known Priorities to OpenFlow priorities.
 	priorityMap map[types.Priority]uint16
-	// priorityOffset stores the current size of a priority zone.
-	// When tiering is introduced, each tier will keep its own priorityOffset.
-	priorityOffset uint16
-	// numPriorityZones stores the current number of numPriorityZones (within the default tier).
-	// When tiering is introduced, each tier will keep its own numPriorityZones.
-	numPriorityZones int32
+	// ofPriorityMap maintains the current mapping of OpenFlow priorities in the table to Priorities.
+	ofPriorityMap map[uint16]types.Priority
+	// sortedPriorities maintains a slice of sorted OpenFlow priorities in the table that are occupied.
+	sortedOFPriorities []uint16
+	// initialOFPriorityFunc determines the initial OpenFlow priority to be checked for input Priorities.
+	initialOFPriorityFunc InitialOFPriorityGetter
 }
 
-func newPriorityAssigner() *priorityAssigner {
+func newPriorityAssigner(initialOFPriorityFunc InitialOFPriorityGetter) *priorityAssigner {
 	pa := &priorityAssigner{
-		priorityMap:      map[types.Priority]uint16{},
-		priorityOffset:   InitialPriorityOffset,
-		numPriorityZones: InitialPriorityZones,
+		priorityMap:           map[types.Priority]uint16{},
+		ofPriorityMap:         map[uint16]types.Priority{},
+		sortedOFPriorities:    []uint16{},
+		initialOFPriorityFunc: initialOFPriorityFunc,
 	}
 	return pa
 }
 
-// getPriorityZoneIndex returns the priorityZone index for the given priority.
-// It maps policyPriority [0.0-1.0) to 0, [1.0-2.0) to 1 and so on so forth.
-// policyPriorities over 99.0 will be mapped to zone 99, without zone expansion for now.
-func (pa *priorityAssigner) getPriorityZoneIndex(p types.Priority) int32 {
-	floorPriority := int32(math.Floor(p.PolicyPriority))
-	if floorPriority > pa.numPriorityZones-1 {
-		floorPriority = pa.numPriorityZones - 1
+// updatePriorityAssignment updates all the local maps to correlate input ofPriority and Priority.
+func (pa *priorityAssigner) updatePriorityAssignment(ofPriority uint16, p types.Priority) {
+	if _, exists := pa.ofPriorityMap[ofPriority]; !exists {
+		priorities := append(pa.sortedOFPriorities, ofPriority)
```

Comment:
Have we considered some sparse bitmap structure? I do not it is available in Golang or not though.

Reply:
The sort priority operation is only called when a new OpenFlow priority from the table is taken by a CNP. When an OpenFlow priority is released, the sorting does not happen again as the priority will just get kicked out of the slice. We need to consider performance improvements for this priorityAssigner in the next release though.

Reply:
When NCP takes a new priority? Every new rule might need that? Then it is still worth optimizations in my mind.

Reply:
Theoretically yes if all new rules are created at a different Tier/policyPriority. But still, the slice only maintains sorted OpenFlow priorities taken from the particular table. What optimization do you have in mind here?

Reply:
How long the slice can be?

I am just thinking a better data structure to search and update the priorities, like:
1. some sorted list.
2. sparse bitmap
3. at least do not sort the whole slice every time - you can just find the right position and insert the new element there (though still need to copy the elements after the new one or sometimes allocate a new slice).

Reply:
I am not sure how sort.Slice is implemented. Maybe it is the same complexity as #3 in your specific case though.

Reply:
Thanks for the suggestion. Digged into Sort package and found out the sort.Slice uses quick sort by default, which is indeed not optimized for out case. I have updated the code above to use insertionSort as you suggested. For the next release will explore bitmap and other plausible solutions

Reply:
i think you can append maybe the last element to increase capacity of slice.. so you can reuse the same slice and then search to find the index and then copying the remainder of the slice till (last-1) and insert the new priority in the index.. i think i read sort.Slice is nlogn

Reply:
> i think you can append maybe the last element to increase capacity of slice.. so you can reuse the same slice and then search to find the index and then copying the remainder of the slice till (last-1) and insert the new priority in the index.. i think i read sort.Slice is nlogn

Did exactly that. Thanks for the suggestion

Reply:
I am ok with the current code. But if we believe it might worth optimization, maybe add a TODO comment here.
Later we can do some performance tests around priority allocation/deallocation?

Reply:
Done.